library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,141 @@
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`
+  - With a rename, the new major version's Java package is `com.google.a2`
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven coordinates (case 1)**: This approach can result in
+    diamond dependency conflicts because different branches of a dependency tree
+    can depend on different major versions, and only one of them can be pulled
+    in by maven in a particular dependency tree. Users are forced to change
+    their code between versions.
+  - **Rename maven coordinates (case 2)**: Users can easily pull in both the old
+    jar (`g1:a1:1.0.0`) and the new jar (`g2:a2:2.0.0`) accidentally through
+    transitive dependencies, because maven treats the two artifacts as distinct,
+    but since they have classes with the same Java packages, they are loaded in
+    unpredictable ways, leading to runtime exceptions (for example
+    `ClassNotFoundException`). It is difficult for users to ensure their build
+    tree only includes one of the artifacts. **NEVER DO THIS**. Example: guava
+    vs guava-jk5.
+- **Rename Java package**:
+  - **Don't rename Maven coordinates (case 3)**: The classes from `g1:a1:1.0.0`
+    and `g1:a1:2.0.0` could technically be used together, but since they share
+    the maven group and artifactId, only one jar can be pulled in. Users are
+    forced to change their code between versions. It is strictly better to also
+    rename maven coordinates, so that the major versions can be used side by
+    side.
+  - **Rename maven coordinates (case 4)**: The two major versions can be used
+    side by side, allowing users to incrementally transition from the old to the
+    new version, or even use them side by side indefinitely if
+    necessary. Transitioning fully to the new version does require code changes
+    between versions, though. There will be no conflicts using this
+    approach. This is the approach taken by Go, described in
+    https://research.swtch.com/vgo-import .
+
+Given the consequences, it seems clear that the two worst options are case 2
+(renaming maven coordinates while keeping the Java package the same) and case 3
+(renaming the Java package while keeping the maven coordinates the same), and
+both should be avoided.
+
+Of the remaining two, case 4 (renaming both maven coordinates and the Java
+package) is the least disruptive and the only option that won't lead to diamond
+dependency conflicts, so it should be the generally preferred approach. However,
+given that a full transition from one major version to the next requires
+updating all import statements regardless of whether the corresponding classes
+have changed, there is potentially a very high cost to this option, which should
+be weighed carefully. Case 1 (neither renaming the maven coordinates nor the
```

Comment:
I feel like case 1 is the default today. Package renames are a mess. I'm not sure we should recommend against as strongly as we do here.